### PR TITLE
Fix slice copy of VersionedSpecs in FeatureGate.

### PR DESCRIPTION
**Pull request by siyuanfoundation**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Address the comments in https://github.com/kubernetes/kubernetes/pull/125941

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/component-base/featuregate/feature_gate.go [hunk 1]
```diff
@@ -152,6 +149,10 @@ type MutableFeatureGate interface {
 	// overriding its default to true for a limited number of components without simultaneously
 	// changing its default for all consuming components.
 	OverrideDefault(name Feature, override bool) error
+	// DeepCopyAndReset copies all the registered features of the FeatureGate object, with all the known features and overrides,
+	// and resets all the enabled status of the new feature gate.
+	// This is useful for creating a new instance of feature gate without inheriting all the enabled configurations of the base feature gate.
+	DeepCopyAndReset() MutableVersionedFeatureGate
```
Comment (sttts) [inline staging/src/k8s.io/component-base/featuregate/feature_gate.go]: a lot better. Thank you!

File: staging/src/k8s.io/component-base/featuregate/feature_gate.go [hunk 2]
```diff
@@ -508,7 +509,9 @@ func (f *featureGate) GetAll() map[Feature]FeatureSpec {
 func (f *featureGate) GetAllVersioned() map[Feature]VersionedSpecs {
 	retval := map[Feature]VersionedSpecs{}
 	for k, v := range f.known.Load().(map[Feature]VersionedSpecs) {
-		retval[k] = v
+		vCopy := make([]FeatureSpec, len(v))
+		_ = copy(vCopy, v)
+		retval[k] = vCopy
```
Comment (sttts) [inline staging/src/k8s.io/component-base/featuregate/feature_gate.go:508]: this is not enough yet. The `Load()` is atomic, but there must be no writers on the slices. I see a couple of methods in here that mutate the slices, e.g. `AddVersioned` and `OverrideDefaultAtVersion`. Those mutations race with the copy here.

Comment (siyuanfoundation) [reply staging/src/k8s.io/component-base/featuregate/feature_gate.go:508]: Thanks for pointing out the nuances! 
I have updated `AddVersioned` and `OverrideDefaultAtVersion` to use a new copy of `map[Feature]VersionedSpecs`. Would that work?

**Comment by fxierh** [comment]
@siyuanfoundation @sttts 

Thanks for the code/review. Let's rebase-merge this so we can trigger presubmit tests from my PR. We can further discuss  from there.

**Pull request merged**